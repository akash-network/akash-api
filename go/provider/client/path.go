package rest

import (
	"fmt"

	mtypes "github.com/akash-network/akash-api/go/node/market/v1beta4"
)

const (
	DeploymentPathPrefix = "/deployment/{dseq}"
	LeasePathPrefix      = "/lease/{dseq}/{gseq}/{oseq}"
	HostnamePrefix       = "/hostname"
	EndpointPrefix       = "/endpoint"
	MigratePathPrefix    = "/migrate"
)

func VersionPath() string {
	return "version"
}

func StatusPath() string {
	return "status"
}

func ValidatePath() string {
	return "validate"
}

func LeasePath(id mtypes.LeaseID) string {
	return fmt.Sprintf("lease/%d/%d/%d", id.DSeq, id.GSeq, id.OSeq)
}

func SubmitManifestPath(dseq uint64) string {
	return fmt.Sprintf("deployment/%d/manifest", dseq)
}

func GetManifestPath(id mtypes.LeaseID) string {
	return fmt.Sprintf("%s/manifest", LeasePath(id))
}

func LeaseStatusPath(id mtypes.LeaseID) string {
	return fmt.Sprintf("%s/status", LeasePath(id))
}

func LeaseShellPath(lID mtypes.LeaseID) string {
	return fmt.Sprintf("%s/shell", LeasePath(lID))
}

func LeaseEventsPath(id mtypes.LeaseID) string {
	return fmt.Sprintf("%s/kubeevents", LeasePath(id))
}

func ServiceStatusPath(id mtypes.LeaseID, service string) string {
	return fmt.Sprintf("%s/service/%s/status", LeasePath(id), service)
}

func ServiceLogsPath(id mtypes.LeaseID) string {
	return fmt.Sprintf("%s/logs", LeasePath(id))
}
