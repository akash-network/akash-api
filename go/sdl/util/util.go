package util

import (
	"math"

	sdk "github.com/cosmos/cosmos-sdk/types"

	rtypes "pkg.akt.dev/go/node/types/resources/v1beta4"
)

func ComputeCommittedResources(factor float64, rv rtypes.ResourceValue) rtypes.ResourceValue {
	// If the value is less than 1, commit the original value. There is no concept of under-commit
	if factor <= 1.0 {
		return rv
	}

	v := rv.Val.Uint64()
	fraction := 1.0 / factor
	committedValue := math.Round(float64(v) * fraction)

	// Don't return a value of zero, since this is used as a resource request
	if committedValue <= 0 {
		committedValue = 1
	}

	result := rtypes.ResourceValue{
		Val: sdk.NewInt(int64(committedValue)),
	}

	return result
}
