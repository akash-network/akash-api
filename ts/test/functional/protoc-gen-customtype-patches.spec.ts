import { describe, expect, it } from "@jest/globals";
import { exec } from "child_process";
import { access, constants as fsConst, readFile, rmdir } from "fs/promises";
import { tmpdir } from "os";
import { join as joinPath } from "path";
import { promisify } from "util";

const execAsync = promisify(exec);

describe("protoc-gen-customtype-patches plugin", () => {
  const config = {
    version: "v2",
    clean: true,
    plugins: [
      {
        local: "ts/script/protoc-gen-customtype-patches.ts",
        strategy: "all",
        out: ".",
        opt: [
          "target=ts",
        ],
      },
    ],
  };

  it("generates `Set` instance with all the types that have reference to fields with custom type option", async () => {
    const outputDir = joinPath(tmpdir(), `ts-bufplugin-${process.pid.toString()}`);
    const protoDir = "./ts/test/functional/proto";
    const command = [
      `buf generate`,
      `--config '${JSON.stringify({
        version: "v2",
        modules: [
          { path: "go/vendor/github.com/cosmos/gogoproto" },
          { path: "./ts/test/functional/proto" },
        ],
      })}'`,
      `--template '${JSON.stringify(config)}'`,
      `-o '${outputDir}'`,
      `--path ${protoDir}/customtype.proto`,
      protoDir,
    ].join(" ");

    try {
      await execAsync(command, {
        cwd: joinPath(__dirname, "..", "..", ".."),
        env: {
          ...process.env,
          BUF_PLUGIN_CUSTOMTYPE_TYPES_PATCHES_OUTPUT_FILE: "customPatches.ts",
        },
      });

      expect(await readFile(joinPath(outputDir, "customPatches.ts"), "utf-8")).toMatchSnapshot();
    } finally {
      if (await access(outputDir, fsConst.W_OK).catch(() => false)) {
        await rmdir(outputDir, { recursive: true });
      }
    }
  });
});
