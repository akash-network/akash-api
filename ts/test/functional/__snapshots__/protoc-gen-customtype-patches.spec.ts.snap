// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`protoc-gen-customtype-patches plugin generates \`Set\` instance with all the types that have reference to fields with custom type option 1`] = `
"import type * as _protos_customtype from "../protos/customtype_pb.none";
import { Dec } from "../../encoding/customTypes/Dec.none";

const t = {
  "akash.test.functional.PetOwner"(value: _protos_customtype.PetOwner, transformType: 'encode' | 'decode') {
    if (value == null) return value;
    const newValue = { ...value };
    if (value.pet != null) newValue.pet = t["akash.test.functional.Pet"](value.pet, transformType);
    return newValue;
  },
  "akash.test.functional.Pet"(value: _protos_customtype.Pet, transformType: 'encode' | 'decode') {
    if (value == null) return value;
    const newValue = { ...value };
    if (value.value?.case === "dog") {
      newValue.dog = {
        ...value.value,
        value: t["akash.test.functional.Dog"](value.value.value, transformType)
      };
    }
    return newValue;
  },
  "akash.test.functional.Dog"(value: _protos_customtype.Dog, transformType: 'encode' | 'decode') {
    if (value == null) return value;
    const newValue = { ...value };
    if (value.name != null) newValue.name = Dec[transformType](value.name);
    return newValue;
  },
  "akash.test.functional.PetClub"(value: _protos_customtype.PetClub, transformType: 'encode' | 'decode') {
    if (value == null) return value;
    const newValue = { ...value };
    if (value.members) newValue.members = value.members.map((item) => t["akash.test.functional.PetOwner"](item, transformType));;
    if (value.allDogs) newValue.allDogs = value.allDogs.map((item) => t["akash.test.functional.Dog"](item, transformType));;
    if (value.ownerToDog) {
      newValue.ownerToDog = {};
      Object.keys(value.ownerToDog).forEach(k => newValue.ownerToDog[k] = t["akash.test.functional.Dog"](value.ownerToDog[k], transformType));
    }
    return newValue;
  },
  "akash.test.functional.PetClubItemsQueryResponse"(value: _protos_customtype.PetClubItemsQueryResponse, transformType: 'encode' | 'decode') {
    if (value == null) return value;
    const newValue = { ...value };
    if (value.items) newValue.items = value.items.map((item) => t["akash.test.functional.PetClub"](item, transformType));
    return newValue;
  },
  "akash.test.functional.Money"(value: _protos_customtype.Money, transformType: 'encode' | 'decode') {
    if (value == null) return value;
    const newValue = { ...value };
    if (value.amount != null) newValue.amount = Dec[transformType](value.amount);
    return newValue;
  }
};

export const transformations = t;
"
`;
